### PR TITLE
Resolve issue 359

### LINKED ISSUE #359: Saving state never terminates if persistent files cannot be accessed / modified
**Describe the bug/problem**
To store data API Dash requires 4 files, namely - 
1. apidash-data.hive
2. apidash-data.lock
3. apidash-settings.hive
4. apidash-settings.lock

Now in Windows, these files are generated inside `%USERPROFILE%/Documents` directory and in Linux, these are generated under `~/` directory. If these locations are protected, i.e. external programs cannot create files in these locations or modify existing files, then if API Dash tries to access these files, it throws an error and the UI never updates. 

**Steps to Reproduce the bug/problem**
Reproducing this issue can be achieved via many ways, i.e. revoking write access to the generated files and trying to save a request. A simpler way to reproduce this issue on Windows is to enable Ransomware Protection which will disallow external programs trying to generate files under some predefined directories (`%USERPROFILE%/Documents` being one of them) and then trying to save a request.

After clicking `save`, the saving state never terminates.

Screenshot 2024-03-25 154145

API Dash throws error in the terminal stating it cannot access the files user to store persistent data.

Screenshot 2024-03-25 153740

**Expected behavior**
User should be presented with an error message stating that the respective files could not be read / modified and file permissions needs to be updated by the user. 

An enhancement to the solution is to add a default location for persistent data and then give the user ability to load data and settings from custom location using file path. This will allow users to version control request templates and share the existing requests among developers.

**Device Info (The device where you encountered this issue):**
 - OS: Windows 11

### PR DESCRIPTION
**Pull request by Sarath191181208**
## Selecting a hive persistence folder 

The flow of the app is as follows:
- on app init the user is prompted to give a save location 
- The folder is picked using `file_picker` package.
- The folder is saved using `shared_preferences`.

**NOTE**: On Linux `file_picker` requires the `zenity` package for picking files.

## Related Issues

- Related Issue #359 
- Closes #359 

### Checklist
- [x] I have gone through the [contributing guide](https://github.com/foss42/apidash/blob/main/CONTRIBUTING.md)
- [x] I have updated my branch and synced it with project `main` branch before making this PR
- [x] I have run the tests (`flutter test`) and all tests are passing, except one test in /apidash/test/models/response_model_test.dart due to invalid SSL certificate of https://api.apidash.dev/

## Added/updated tests?
_We encourage you to add relevant test cases._

- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by ashitaprasad** [comment]
Check `file_selector` we already use this package for picking files in API Dash and it does not requires the zenity package for picking files on Linux.

**Comment by Sarath191181208** [comment]
I am extremely sorry if this is because of me using arch but it requires `zenity` a system package for picking files and gives this error if zenity isn't found on path while giving this error: 
```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Exception: Couldn't find the executable zenity in the path.
#0      isExecutableOnPath (package:file_picker/src/utils.dart:60:5)
<asynchronous suspension>
#1      FilePickerLinux._getPathToExecutable (package:file_picker/src/linux/file_picker_linux.dart:123:14)
<asynchronous suspension>
#2      FilePickerLinux.getDirectoryPath (package:file_picker/src/linux/file_picker_linux.dart:67:24)
<asynchronous suspension>
#3      HiveDirectorySelectorState.selectFolder (package:apidash/widgets/hive_directory_selector.dart:18:33)
<asynchronous suspension>
```

**Comment by ashitaprasad** [comment]
Did you read my comment above?

**Comment by ashitaprasad** [comment]
@Sarath191181208 is this PR ready for review?

**Comment by Sarath191181208** [comment]
Yes

File: lib/models/hive_storage_folder.dart
Comment (animator) [inline lib/models/hive_storage_folder.dart:1]: These are not models, but utilities. Should be in the right location.

File: lib/extensions/load_hive_extension.dart [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import 'package:hive_flutter/hive_flutter.dart';
+import 'package:path/path.dart';
+
+/// Flutter extensions for Hive.
+extension HiveWithFolder on HiveInterface {
+  /// Initializes Hive with the path from [appDir].
+  ///
+  /// You can provide a [subDir] where the boxes should be stored.
+  Future<void> initHiveWithRootFolder(String appDir, [String? subDir]) async {
+    WidgetsFlutterBinding.ensureInitialized();
```
Comment (ashitaprasad) [inline lib/extensions/load_hive_extension.dart:12]: WidgetsFlutterBinding.ensureInitialized(); should not be hidden in this function. It should be in `main()` as it is currently.

File: lib/extensions/load_hive_extension.dart [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    if (kIsWeb) return;
+
+    init(join(appDir, subDir));
+  }
+}
```
Comment (ashitaprasad) [inline lib/extensions/load_hive_extension.dart:17]: What is the purpose of this extension. This task can be done by a hive utility function.

File: lib/models/hive_storage_folder.dart [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Store hive save folder path.
```
Comment (ashitaprasad) [inline lib/models/hive_storage_folder.dart:1]: These are not models. They are hive utility functions.

File: lib/services/hive_services.dart [hunk 4]
```diff
@@ -8,8 +9,8 @@ const String kKeyEnvironmentBoxIds = "environmentIds";
 
 const String kSettingsBox = "apidash-settings";
 
-Future<void> openBoxes() async {
-  await Hive.initFlutter();
+Future<void> openBoxes(String hiveSaveFolder) async {
+  await Hive.initHiveWithRootFolder(hiveSaveFolder);
```
Comment (ashitaprasad) [inline lib/services/hive_services.dart:13]: Will get changed according to the comment above

File: lib/widgets/hive_directory_selector.dart [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import 'package:apidash/models/hive_storage_folder.dart';
+import 'package:apidash/services/hive_services.dart';
+import 'package:flutter/material.dart';
+
+class HiveDirectorySelector extends StatefulWidget {
```
Comment (ashitaprasad) [inline lib/widgets/hive_directory_selector.dart:5]: Does it require a Stateful Widget or Stateless can work?

**Pull request closed without merging**